text/2021-10-08-monitoring-metrics-about-chaos-mesh.md

```diff
@@ -0,0 +1,241 @@
+# Monitoring Metrics about Chaos Mesh
+
+## Summary
+
+More metrics is needed for improving observability of `chaos-controller-manager`,
+`chaos-daemon` and `chaos-dashboard`. Now, we already have several metrics in
+`chaos-controller-manager` and `chaos-daemon`, but it's not enough.
+
+## Motivation
+
+At present, we only collected several metrics about webhooks like
+`chaos_mesh_injections_total`, chaos experiments information like
+`chaos-controller-manager`, and HTTP and gRPC in `chaos-daemon`.
+These metrics are not easy to reflect the overall appearance of chaos-mesh system,
+so we need more metrics to improve the observability.
+
+As for the expected outcome, below is several metrics about logic pattern and
+performance should be implemented and exposed on `/metrics` HTTP endpoint:
+
+chaos-controller-manager:
+
+- time histogram for each `Reconcile()` in `Reconciler`
+- count of Chaos Experiments, Schedule, and Workflow
+- count of the emitted event by chaos-controller-manager
+- common metrics of grpc-client
+- metrics for kubernetes webhook
+
+chaos-daemon:
+
+- common metrics of grpc-server (already provided)
+- count of processes controlled by `bpm` (background process manager)
+- count of iptables/ipset/tc rules
+
+chaos-dashboard:
+
+- time histogram for each HTTP query
+- count for archived object
+- time histogram for archive reconciler
+
+## Detailed design
+
+### Metrics Plan
+
+The metrics design is as follows. It should be noted that the implementations
```

Comment:
Can we add some metrics to monitor the number and status of chaos-daemon component? In addition, the metric about the number of GRPC errors is also very useful.

Reply:
> Can we add some metrics to monitor the number and status of chaos-daemon component?

If I understand correctly, we want to monitor the number of available chaos-daemon. We do NOT need to build these metrics somewhere "on purpose". We could do that with Prometheus service discovery with kubernetes, prometheus will create metrics scrape job for each chaos-daemon instance(with the different value in `job` tag), we could use a simple `count()` to calculating the current available chaos-daemon instane.